Editors/PathEditor.cpp

```diff
@@ -1,10 +1,238 @@
+
+  ui->statusBar->addWidget(cursorPositionLabel);
+
+  ui->pointsTableView->installEventFilter(this);
+
+  connect(xSnap, QOverload<int>::of(&QSpinBox::valueChanged),
+          [this](int val) { ui->roomView->GetGrid().horSpacing = val; });
+
+  connect(ySnap, QOverload<int>::of(&QSpinBox::valueChanged),
+          [this](int val) { ui->roomView->GetGrid().vertSpacing = val; });
+
+  connect(this, &BaseEditor::FocusGained, [=]() { ui->pathPreviewBackground->parentHasFocus = true; });
+  connect(this, &BaseEditor::FocusLost, [=]() { ui->pathPreviewBackground->parentHasFocus = false; });
+
+  nodeMapper->addMapping(ui->nameEdit, TreeNode::kNameFieldNumber);
+
+  resMapper->addMapping(ui->smoothCheckBox, Path::kSmoothFieldNumber);
+  resMapper->addMapping(ui->closedCheckBox, Path::kClosedFieldNumber);
+  resMapper->addMapping(ui->precisionSpinBox, Path::kPrecisionFieldNumber);
+  resMapper->addMapping(xSnap, Path::kHsnapFieldNumber);
+  resMapper->addMapping(ySnap, Path::kVsnapFieldNumber);
+
+  RebindSubModels();
 }
 
 PathEditor::~PathEditor() { delete ui; }
+
+void PathEditor::RebindSubModels() {
+  pathModel = _model->GetSubModel(TreeNode::kPathFieldNumber);
+
+  ui->roomView->SetPathModel(pathModel);
+  pointsModel = pathModel->GetRepeatedSubModel(Path::kPointsFieldNumber);
+  ui->pointsTableView->setModel(pointsModel);
+  ui->pointsTableView->hideColumn(0);
+
+  QString roomName = pathModel->data(Path::kBackgroundRoomNameFieldNumber).toString();
+  if (roomName != "") {
+    ui->roomView->SetResourceModel(
+        MainWindow::resourceMap->GetResourceByName(TypeCase::kRoom, roomName)->GetSubModel(TreeNode::kRoomFieldNumber));
+  }
+  roomLineEdit->setText(roomName);
+
+  connect(ui->pathPreviewBackground, &AssetScrollAreaBackground::MouseMoved, [=](int x, int y) {
+    const GridDimensions g = ui->roomView->GetGrid();
+    QPoint mousePos(x, y);
+    if (snapToGrid) mousePos = QPoint(RoundNum(x, g.horSpacing), RoundNum(y, g.vertSpacing));
+    ui->roomView->mousePos = mousePos;
+    cursorPositionLabel->setText(tr("X %0, Y %1").arg(mousePos.x()).arg(mousePos.y()));
+    if (draggingPoint) {
+      pointsModel->setData(ui->roomView->selectedPointIndex, Path::Point::kXFieldNumber, ui->roomView->mousePos.x());
+      pointsModel->setData(ui->roomView->selectedPointIndex, Path::Point::kYFieldNumber, ui->roomView->mousePos.y());
+    } else
+      ui->pathPreviewBackground->update();  // manually call update to redraw cursor if no data changes
+  });
+
+  connect(ui->pathPreviewBackground, &AssetScrollAreaBackground::MousePressed, [=](Qt::MouseButton button) {
+    if (button == Qt::MouseButton::LeftButton) {
+      draggingPoint = true;
+      // Check for point at location (reverse loop to get last added one if dups)
+      for (int i = pointsModel->rowCount() - 1; i >= 0; i--) {
+        QPoint pt(pointsModel->data(i, Path::Point::kXFieldNumber).toInt(),
+                  pointsModel->data(i, Path::Point::kYFieldNumber).toInt());

```

Comment:
There's a Point() accessor on PathView for this, now.

Reply:
Ok whatever. I'll use it